src/main/capi/arrow-c.cpp

```diff
@@ -14,6 +16,177 @@ using duckdb::PreparedStatementWrapper;
+	if (!client_properties || !out_arrow_array || !chunk) {
+		return duckdb_create_error_data(DUCKDB_ERROR_INVALID_INPUT,
+		                                "Invalid argument(s) to duckdb_data_chunk_to_arrow");
+	}
+	auto dchunk = reinterpret_cast<duckdb::DataChunk *>(chunk);
+	auto client_properties_wrapper = reinterpret_cast<CClientPropertiesWrapper *>(*client_properties);
+	auto extension_type_cast = duckdb::ArrowTypeExtensionData::GetExtensionTypes(
+	    *client_properties_wrapper->properties.client_context, dchunk->GetTypes());
+
+	try {
+		ArrowConverter::ToArrowArray(*dchunk, reinterpret_cast<ArrowArray *>(*out_arrow_array),
+		                             client_properties_wrapper->properties, extension_type_cast);
+	} catch (const duckdb::Exception &ex) {
+		return duckdb_create_error_data(DUCKDB_ERROR_INVALID_INPUT, ex.what());
+	} catch (const std::exception &ex) {
+		return duckdb_create_error_data(DUCKDB_ERROR_INVALID_INPUT, ex.what());
+	} catch (...) {
+		return duckdb_create_error_data(DUCKDB_ERROR_INVALID_INPUT, "Unknown error occurred during conversion");
+	}
+	return nullptr;
+}
+
+duckdb_error_data arrow_to_duckdb_schema(duckdb_connection connection, duckdb_arrow_schema schema,
+                                         duckdb_arrow_converted_schema *out_types, char ***out_names,
+                                         idx_t *out_column_count) {
+	if (!connection || !schema || !out_types || !out_names || !out_column_count) {
+		return duckdb_create_error_data(DUCKDB_ERROR_INVALID_INPUT,
+		                                "Invalid argument(s) to duckdb_data_chunk_to_arrow");
+	}
+	duckdb::vector<std::string> names;
+	auto conn = reinterpret_cast<Connection *>(connection);
+	auto arrow_table = new duckdb::ArrowTableType();
+	*out_types = reinterpret_cast<duckdb_arrow_converted_schema>(arrow_table);
+	try {
+		duckdb::vector<LogicalType> return_types;
+		duckdb::ArrowTableFunction::PopulateArrowTableType(duckdb::DBConfig::GetConfig(*conn->context), *arrow_table,
+		                                                   *reinterpret_cast<ArrowSchema *>(schema), names,
+		                                                   return_types);
+		QueryResult::DeduplicateColumns(names);
+	} catch (const duckdb::Exception &ex) {
+		delete arrow_table;
+		return duckdb_create_error_data(DUCKDB_ERROR_INVALID_INPUT, ex.what());
+	} catch (const std::exception &ex) {
+		delete arrow_table;
+		return duckdb_create_error_data(DUCKDB_ERROR_INVALID_INPUT, ex.what());
+	} catch (...) {
+		delete arrow_table;
+		return duckdb_create_error_data(DUCKDB_ERROR_INVALID_INPUT, "Unknown error occurred during conversion");
+	}
+	const idx_t column_count = names.size();
+	*out_column_count = column_count;
+	*out_names = new char *[column_count];
+	for (idx_t i = 0; i < column_count; i++) {
+		(*out_names)[i] = new char[names[i].size() + 1];
+		std::strcpy((*out_names)[i], names[i].c_str());
```

Comment:
I'm not sure these are null-terminated, and neither is the outer `out_names[column_count]`

Reply:
`names[i].c_str()` returns a null-terminated c-string, so it should be null-terminated.
But I think you are right for `out_names[column_count]`.

Reply:
Ah yes, I'm mistaken, I just checked the manpage:
`The stpcpy() and strcpy() functions copy the string src to dst (including the terminating ‘\0’ character.)`

Seems like the null terminator is included